### DISCUSSION TITLE
Bug initing speechly from a robot server

### DISCUSSION THREAD

**Discussion by cfry** [speechly/speechly#277] [Q&A]
I have a complex web app that drives a robot via speech.
It uses fairly vanilla JavaScript and HTML.

I have a local server serving this app.
I wrote the server, but the server has zero code in it related to Speechly.
When I run the app with the server on my Mac laptop,
it serves the app to Chrome (on the same Mac laptop) 
and speech reco works great.

The robot has an ARM processor running linux
and the same server code that I run on my Mac.
When I run the server on the robot
and I use Chrome on my laptop to browse the web app served by the robot,
the UI pops up ok, but when
I attempt to initialize the mic, I get the error:
"Cannot read properties of undefined (reading 'getSupportedConstraints')"

My method that is called when this error occurs is:

```
static async init_speechly(){
        if(this.speech_reco_possible) {
            try {
                this.client = new BrowserClient({
                    appId: 'my app id', //the real code has the real app id
                    vad: {enabled: true, //enables voice activated detection, no need to use stop and start except to pause
                          noiseGateDb: -24.0} // https://github.com/speechly/speechly/blob/main/libraries/browser-client/docs/interfaces/client.VadOptions.md#enabled
                })
                this.is_vad_enabled = true
                this.microphone = new BrowserMicrophone()
                // React to the updates from the API.
                this.client.onSegmentChange((segment) => {
                    console.log('Received new segment:', segment
                      
                    ) //the duration between getting a non-final "stop" and a final "go" is
                    // 20ms??? and undetetable by me (the dur for both is about 500ms)
                    //so might as well not do special processing and send "stop" thru the regular pipeline
                    
                    if (segment.isFinal) {
                        console.log('Received final segment of: ' + this.segment_to_text(segment))
                        Talk.handle_command(segment)
                    }
                })
                // Initialize the microphone - this will ask the user for microphone permissions
// and establish the connection to Talk API.
// Make sure you call `initialize` from a user action handler
// (e.g. from a button press handler).
                await this.microphone.initialize()

// bind the microphone to the client
                await this.client.attach(this.microphone.mediaStream)// Initialize the microphone - this will ask the user for microphone permissions
                //this.client.start() //done only in turn_on_mic
            }
            catch(err){
                warning("Sorry, can't connect to internet-based speech recognition.<br/>" +
                         "Check your connectivity.<br/>" +
                         err.message)
                this.speech_reco_possible = false
            }
        }
    }
```

Something in that TRY code is causing the CATCH clause to be invoked.
I'm guessing:
```
new BrowserClient(...)
new BrowserMicrophone()
this.microphone.initialize()
or
this.client.attach(this.microphone.mediaStream)
```

The chrome debugger breaks in your file speechly.es.js
in the following code that starts on line 1706

```
this.speechlyNode.port.onmessage = (event) => {
                    switch (event.data.type) {
                        case 'STATS':
                            if (event.data.signalEnergy > this.stats.maxSignalEnergy) {
                                this.stats.maxSignalEnergy = event.data.signalEnergy;
                            }
                            this.stats.sentSamples += parseInt(event.data.samples);
                            break;
                        case 'DATA':
                            // this is not called if SAB is used, the buffers are sent immediately
                            this.handleAudio(event.data.frames);
                            break;
                    }
                };
```
The line it breaks on is:
`this.stats.maxSignalEnergy = event.data.signalEnergy`
Before this line is executed,
`this.stats.maxSignalEnergy` is bound to 0
event is bound, and has about 50 properties, all look like normal
JS event properties, but "data" is not one of the properties.

Ideas greatly appreciated.

**Comment by nameisxi** [comment]
Hi @cfry, thanks for reaching out!

Did I understand correctly that you’re running a web app (backend & frontend) locally from your Mac and that you then use the frontend of the web app to send commands to the robot? In addition to that, the same backend is also running on the robot, but doesn’t have any Speechly-related code?

Here’s my understanding of the architecture of your project:
[ Backend (running on Mac) ] <-> [ Frontend (running on Mac/Chrome. Has all the Speechly-related code) ] <-> [ Backend (running on the robot) ]

Let me know if I have misunderstood something.

Could you also reply to the following questions:
- Which Mac do you have?
- Which MacOs version are you running?
- Which Chrome version are you running?

Thank you!

**Reply by nameisxi** [reply]
Also, have you had a chance to try the solution described [here](https://stackoverflow.com/a/70151770)?

**Comment by cfry** [comment]
Tuomos,
thanks so much for your response.
In your first message, no you aren't understanding my architecture
so I guess I did a bad job of explaining it
BUTTTTT
your 2nd message about the stack overflow has a clue that looks quite
valuable
as I am seeing, like that stack overflow thread indicates,
"Cannot read properties of undefined (reading 'getSupportedConstraints')"
so I think you're on to something.
Following that thread, it suggest I
browse
chrome://flags/#unsafely-treat-insecure-origin-as-secure
and enter in a url.
The question is, what URL?
From Speechly in my web page, as far as I can tell,
I'm only using
new BrowserClient()   and
new BrowserMicrophone()
neither of which take a url as an argument.
SO I'm guessing there's some URL inside those functions that
is used to transmit the audio to Speechly in the cloud
and get back the response???
Does this URL include my appID?

**Reply by cfry** [reply]
Sorry for my late response. Its hard for me to directly test this.
From your response above:
"Did I understand correctly that you’re running a web app (backend & frontend) locally from your Mac"
yes and that's the scenario that works.
My 2nd scenario is that
we run the backend served from a robot and the front end is the same as scenario #1, ie chrome running
on my Mac hitting the server on the robot. This scenario does not work.
The first time I did this the robot was NOT connected to the net.
But with some difficulty I was able to connect it to the net and it also DIDN'T work.

Also, doing the trick you suggested in the stack overflow sounded promising because I was
seeing the same kind of error, ie:
TypeError: Cannot read property 'getSupportedConstraints' of undefined.
So I tried setting the chrome flag:
chrome://flags/#unsafely-treat-insecure-origin-as-secure
and inserting the url's of:
api.speechly.com or 
api.speechly.com/ws or
wss://api.speechly.com
 and none of them changed the bad behavior.
(am I using the wrong URL here?)

One idea is that my Mac (the one I signed up for the Speechly account, and the one
where everything works as long as I'm serving the web page that calls the 
Speechly api for Browser and Microphone,)
has some special cookie, or browser local storage permission or something installed by
your software that prevents other users/computers from using my Speechly ID 
from working.
Am I suppose to be able to let other people use this code from their own browsers using my Speechly ID?

PS:
Which Mac do you have? MacBookPro 15 inch, circa 2019
Which MacOs version are you running? Ventura 13.2
Which Chrome version are you running? 
"Chrome is up to date
Version 110.0.5481.177 "

**Reply by cfry** [reply]
I've got a little more info.
I was able to get the core of my code running on a Windows machine with a local server on
the windows machine and chrome browser running on the same windows machine,
using my same Speechly apiId, so that's good.  (and the accuracy of the speech reco was excellent!) 
In my case that doesn't work, the laptop running chrome is a linux machine.
Should that be ok?

**Reply by nameisxi** [reply]
Are you using HTTPS when making these requests? [It seems](https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/getUserMedia#privacy_and_security) that non-HTTPS and file:/// requests make the `getUserMedia()` function return undefined.

**Comment by cfry** [comment]
Sorry I forgot to put this in my previous post.
I'm not making any requests in my code.
Speechly BrowserClient and/or BrowserMicrophone
must be doing it, so you tell me what the protocol you're using is.
I'm using
vad enabled: true
in case that matters.

Now when I followed your suggestion of that stackoverflow
issue you mentioned that said modify
chrome://flags/#unsafely-treat-insecure-origin-as-secure
I asked you what URL I should use as its value (you didn't respond)
I tried:
wss://api.speechly.com
because that's what Chrome Dev tools indicated was being requested.
I also tried:
 api.speechly.com
api.speechly.com/ws
http:api.speechly.com
https:api.speechly.com
Please tell me what I should use.

Also, is there any reason that running the speechly client code
in Chrome on a Linux machine wouldn't work?

**Reply by nameisxi** [reply]
I think that the problem lies in how the requests are being made. If you're doing this in localhost, it's an unsecure origin, and needs to be setup manually in Chrome ([see instructions](https://medium.com/@Carmichaelize/enabling-the-microphone-camera-in-chrome-for-local-unsecure-origins-9c90c3149339)). This shouldn't be an issue in production, assuming you have setup HTTPS.
